core/policy/match.go

```diff
@@ -0,0 +1,162 @@
+package policy
+
+import (
+	"cmp"
+	"fmt"
+
+	"github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/must"
+	"github.com/storacha-network/go-ucanto/core/policy/selector"
+)
+
+// Match determines if the IPLD node matches the policy document.
+func Match(policy Policy, node ipld.Node) bool {
+	for _, stmt := range policy {
+		ok := matchStatement(stmt, node)
+		if !ok {
+			return false
+		}
+	}
+	return true
+}
+
+func matchStatement(statement Statement, node ipld.Node) bool {
+	switch statement.Kind() {
```

Comment:
does it not make sense to make this a type switch? https://go.dev/tour/methods/16

asking cause it's a private method, you have access to the underlying types.

also, do you expect others outside the policy module to implement the statement interfaces? If not, I'd just skip them. several are also identical. also if it's a closed union, you can just make a private interface method all the underlying types implement, and since it's private the interface won't be implementable outside the module. Here's golang's own ast parser doing it... https://cs.opensource.google/go/go/+/refs/tags/go1.20.4:src/go/ast/ast.go;l=851

Reply:
basically,

interface with private method and type switch is my go to for tagged unions, which don't really exist in go. (there's some debate about other solutions like visitor pattern, but I find them too verbose)

Reply:
FYI all the policy code got moved into `go-ucan` and I believe they changed it to something similar to what you suggested :)